File: aihound/utils/keychain.py
"""macOS Keychain query utilities."""

from __future__ import annotations

import subprocess
from typing import Optional

from aihound.core.platform import detect_platform, Platform


def query_keychain(service: str) -> Optional[str]:
    """Query macOS Keychain for a credential by service name.

    Returns the password/token value if found, None otherwise.
    Only works on macOS.
    """
    if detect_platform() != Platform.MACOS:
        return None

    try:
        result = subprocess.run(
            [
                "security", "find-generic-password",
                "-s", service,
                "-w",  # Output only the password
            ],
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            return result.stdout.strip()
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        pass

    return None


def list_keychain_entries(service_filter: str = "") -> list[dict]:
    """List Keychain entries, optionally filtered by service name.

    Returns list of dicts with 'service', 'account' keys.
    Only works on macOS.
    """
    if detect_platform() != Platform.MACOS:
        return []

    try:
        result = subprocess.run(
            ["security", "dump-keychain"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode != 0:
            return []

        entries = []
        current = {}
        for line in result.stdout.splitlines():
            line = line.strip()
            if line.startswith('"svce"'):
                # Service name
                val = _extract_keychain_value(line)
                if val:
                    current["service"] = val
            elif line.startswith('"acct"'):
                val = _extract_keychain_value(line)
                if val:
                    current["account"] = val
            elif line == "attributes:":
                if current and (not service_filter or service_filter in current.get("service", "")):
                    entries.append(current)
                current = {}

        if current and (not service_filter or service_filter in current.get("service", "")):
            entries.append(current)

        return entries

    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        return []


def _extract_keychain_value(line: str) -> Optional[str]:
    """Extract value from a Keychain dump line like '"svce"<blob>="My Service"'."""
    if '="' in line:
        _, _, val = line.partition('="')
        return val.rstrip('"')
    return None

File: Other Versions/Go/utils/keychain.go
// Package utils provides platform-specific credential access utilities.
package utils

import (
	"context"
	"os/exec"
	"strings"
	"time"

	"aihound/core"
)

// KeychainEntry represents a single entry from the macOS Keychain.
type KeychainEntry struct {
	Service string
	Account string
}

// QueryKeychain queries the macOS Keychain for a credential by service name.
// Returns the password/token value if found, or an error otherwise.
// Only works on macOS.
func QueryKeychain(service string) (string, error) {
	if core.DetectPlatform() != core.PlatformMacOS {
		return "", ErrNotMacOS
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "security", "find-generic-password", "-s", service, "-w")
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return strings.TrimSpace(string(output)), nil
}

// ListKeychainEntries lists Keychain entries, optionally filtered by service name.
// Only works on macOS.
func ListKeychainEntries(serviceFilter string) ([]KeychainEntry, error) {
	if core.DetectPlatform() != core.PlatformMacOS {
		return nil, ErrNotMacOS
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "security", "dump-keychain")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var entries []KeychainEntry
	var current KeychainEntry
	hasData := false

	for _, line := range strings.Split(string(output), "\n") {
		line = strings.TrimSpace(line)

		if strings.HasPrefix(line, `"svce"`) {
			if val := extractKeychainValue(line); val != "" {
				current.Service = val
				hasData = true
			}
		} else if strings.HasPrefix(line, `"acct"`) {
			if val := extractKeychainValue(line); val != "" {
				current.Account = val
				hasData = true
			}
		} else if line == "attributes:" {
			if hasData && (serviceFilter == "" || strings.Contains(current.Service, serviceFilter)) {
				entries = append(entries, current)
			}
			current = KeychainEntry{}
			hasData = false
		}
	}

	// Don't forget the last entry.
	if hasData && (serviceFilter == "" || strings.Contains(current.Service, serviceFilter)) {
		entries = append(entries, current)
	}

	return entries, nil
}

// extractKeychainValue extracts the value from a Keychain dump line such as
// "svce"<blob>="My Service".
func extractKeychainValue(line string) string {
	idx := strings.Index(line, `="`)
	if idx < 0 {
		return ""
	}
	val := line[idx+2:]
	return strings.TrimRight(val, `"`)
}
